woocommerce/woocommerce-gutenberg-products-block
# Fix: Product Collection show products with correct stock statuses

## What

This PR fixes the issue that Product Collection shows out of stock products event if that status isn't selected:

1. Set some product to be out of stock.
2. Add Product Collection block to a page.
3. Set the stock status filter to all options except `outofstock`.
4. Still see the out of stock products in the editor preview.
5. Save and see the same on the front end. Out of stock products still show up.

Fixes #

## Why

We're handling query related to stock status a little more complex than other:
- If the selected stock statuses contain all available statuses or no status was selected, we do nothing.
- If the selected stock statuses contain all available statuses **but not `outofstock`**, we add out of stock term to product visibility tax query.
- For the remaining cases, we use a meta query.

We do it for some reasons:
- Following WC Core.
- Performance: using tax queries is more performant than using meta queries.



## Testing Instructions



_Please consider any edge cases this change may have, and also other areas of the product this may impact._

1. Set some product to be out of stock.
2. Add Product Collection block to a page.
3. Set the stock status filter to all options except `outofstock`.
4. See the out of stock products **not show up** in the editor preview.
5. Save and see the same on the front end. Out of stock products don't show up.


## WooCommerce Visibility


Required:


## Checklist

Required:

Conditional:

## Changelog


> Fix: Product Collection show products with correct stock statuses
